File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int readyTruckCnt = truck_weights.length;
        int complitTruckCnt = 0;
        int truck_index = 0;
        
        Queue<Integer> road = new LinkedList<>();
        road.offer(0);
        int roadWeight = 0;
        
        while(!road.isEmpty()) {
            // 트럭이 있을 때
            if (truck_index < truck_weights.length) {
                int truck = truck_weights[truck_index];

                
                // 트럭을 road에 넣기
                if (roadWeight + truck <= weight) {
                    road.offer(truck);
                    roadWeight += truck;
                    truck_index++;
                } else {
                    road.offer(0);
                }
                
                // 트럭 빼기
                if (road.size() >= bridge_length) {
                    int complitTruck = road.poll();
                    roadWeight -= complitTruck;
                }
                answer++;
            } 
            // 트럭이 이미 다리에 올라간 경우
            else {
                answer += bridge_length;
                break;
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
# 다리를 건너는 시간을 구하고자한다.
# 다리는 최대 bridge_length대 만큼만 올라갈 수 있다
# 다리는 weight 이하의 무게를 견딜 수 있다

from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    ready = deque(truck_weights)
    bridge = deque()
    
    t_count = 0
    t_weight = 0
    while ready:
        answer += 1
        truck = ready[0]
        if t_count < bridge_length and t_weight + truck <= weight:
            ready.popleft()
            t_count += 1
            t_weight += truck
            bridge.append([truck, 0])
        
        for i in range(len(bridge)):
            bridge[i][1] += 1

        if bridge[0][1] >= bridge_length:
            finish = bridge.popleft()
            t_count -= 1
            t_weight -= finish[0]
            
    answer += bridge_length    
        
    return answer

